File: pyfenstein3d/engine/map2d.py
from .item import Item
from .wall import Wall
from .decoration import Decoration
from .item_grid import ItemGrid
from .player import Player
from .person import Person
from .door import Door
from .item_weapon import ItemWeapon
from .item_ammo import ItemAmmo
from .item_score import ItemScore


class Map2d():
    def __init__(self, items: []):
        self.__grid = ItemGrid(items)
        self.__players = {item.player_id: item for item in items if isinstance(item, Player)}
        self.__persons = [item for item in items if isinstance(item, Person)]
        self.__doors = [item for item in items if isinstance(item, Door)]

    @property
    def grid(self) -> Item:
        return self.__grid

    def update(self, delta_time: float):
        for player_id in self.__players:
            player = self.__players[player_id]
            player.update(delta_time, self.__grid)
            player.adjust_collision(self.__grid)
            player.cast(self.__grid)
        for door in self.__doors:
            door.update(delta_time, self.__persons)


    def get_player(self, player_id: str):
        return self.__players[player_id]

    @staticmethod
    def create_with_pattern(pattern: str):
        items = []
        lines = pattern.splitlines()
        list_type_ids = [[s[n:n+2]
                          for n in range(0, len(s), 2)] for s in lines]
        for block_y, type_ids in enumerate(list_type_ids):
            for block_x, type_id_hex in enumerate(type_ids):
                if type_id_hex.strip() == "":
                    continue
                type_id = int(type_id_hex, 16)
                if type_id in type_ids_walls_solid:
                    items.append(Wall(block_x, block_y, type_id))
                elif type_id in type_ids_weapon:
                    items.append(ItemWeapon(block_x + 0.5, block_y + 0.5, type_id, False))
                elif type_id in type_ids_ammo:
                    items.append(ItemAmmo(block_x + 0.5, block_y + 0.5, type_id, False))
                elif type_id in type_ids_score:
                    items.append(ItemScore(block_x + 0.5, block_y + 0.5, type_id, False))
                elif type_id in type_ids_decorations_non_solid:
                    items.append(Decoration(block_x + 0.5, block_y + 0.5, type_id, False))
                elif type_id in type_ids_decorations_solid:
                    items.append(Decoration(block_x + 0.5, block_y + 0.5, type_id, True))
                elif type_id in type_ids_doors_horizontal:
                    items.append(Door(block_x, block_y + 0.5, type_id, False))
                elif type_id in type_ids_doors_vertical:
                    items.append(Door(block_x + 0.5, block_y, type_id, True))
                elif type_id in type_ids_player:
                    items.append(Player.create(block_x, block_y, type_id))
        return Map2d(items)



type_ids_decorations_solid = [57, 60, 61, 63, 65, 66, 68,
                         69, 70, 71, 74, 75, 76, 80, 93, 94, 95, 97]
# type_ids_decorations_non_solid = [58, 62, 67,
#                              72, 73, 77, 81, 92, 96, 99, 100, 101, 102]

type_ids_decorations_non_solid = [i for i in range(56,120) if i not in type_ids_decorations_solid]
type_ids_walls_solid = list(range(48)) + [53, 54]
type_ids_score = [ 87, 88, 89 ]
type_ids_weapon = [ 85, 86 ]
type_ids_ammo = [ 84 ]

type_ids_doors_horizontal = [50]
type_ids_doors_vertical = [49]
type_ids_player = [255]

File: go/engine/map2d.go
package engine

import (
	"strconv"
	"strings"
)

// Map2d holds a parsed game map with all its entities.
type Map2d struct {
	grid    *ItemGrid
	players map[string]*Player
	persons []Blocker
	doors   []*Door
	enemies []*Enemy
}

// NewMap2d creates a Map2d from a list of blocks.
func NewMap2d(items []Blocker) *Map2d {
	m := &Map2d{
		grid:    NewItemGrid(items),
		players: make(map[string]*Player),
	}
	for _, item := range items {
		if p, ok := item.(*Player); ok {
			m.players[p.PlayerID()] = p
			m.persons = append(m.persons, item)
		}
		if d, ok := item.(*Door); ok {
			m.doors = append(m.doors, d)
		}
		if e, ok := item.(*Enemy); ok {
			m.enemies = append(m.enemies, e)
			m.persons = append(m.persons, item)
		}
	}
	return m
}

func (m *Map2d) Grid() *ItemGrid { return m.grid }

// Update advances all players and doors by deltaTime.
func (m *Map2d) Update(deltaTime float64) {
	for _, player := range m.players {
		player.Update(deltaTime, m.grid)
		player.AdjustCollision(m.grid)
		player.Cast(m.grid)
	}
	var persons []Blocker
	for _, player := range m.players {
		persons = append(persons, player)
	}
	for _, door := range m.doors {
		door.Update(deltaTime, persons)
	}
}

// GetPlayer returns the player with the given ID.
func (m *Map2d) GetPlayer(playerID string) *Player {
	return m.players[playerID]
}

// CreateWithPattern parses a hex pattern string and returns a Map2d.
func CreateWithPattern(pattern string) *Map2d {
	var items []Blocker
	lines := strings.Split(pattern, "\n")
	for blockY, line := range lines {
		for n := 0; n+1 < len(line); n += 2 {
			chunk := line[n : n+2]
			if strings.TrimSpace(chunk) == "" {
				continue
			}
			val, err := strconv.ParseInt(chunk, 16, 32)
			if err != nil {
				continue
			}
			typeID := int(val)
			blockX := n / 2
			bxf := float64(blockX)
			byf := float64(blockY)

			switch {
			case isWallSolid(typeID):
				items = append(items, NewWall(bxf, byf, typeID))
			case isWeapon(typeID):
				items = append(items, NewItemWeapon(bxf+0.5, byf+0.5, typeID))
			case isAmmo(typeID):
				items = append(items, NewItemAmmo(bxf+0.5, byf+0.5, typeID))
			case isScore(typeID):
				items = append(items, NewItemScore(bxf+0.5, byf+0.5, typeID))
			case isDecorationNonSolid(typeID):
				items = append(items, NewDecoration(bxf+0.5, byf+0.5, typeID, false))
			case isDecorationSolid(typeID):
				items = append(items, NewDecoration(bxf+0.5, byf+0.5, typeID, true))
			case isDoorHorizontal(typeID):
				items = append(items, NewDoor(bxf, byf+0.5, typeID, false))
			case isDoorVertical(typeID):
				items = append(items, NewDoor(bxf+0.5, byf, typeID, true))
			case isEnemy(typeID):
				items = append(items, CreateEnemy(bxf+0.5, byf+0.5, typeID))
			case isPlayerType(typeID):
				items = append(items, CreatePlayer(bxf+0.5, byf+0.5, typeID))
			}
		}
	}
	return NewMap2d(items)
}

var typeIDsDecorationsSolid = map[int]bool{
	57: true, 60: true, 61: true, 63: true, 65: true, 66: true, 68: true,
	69: true, 70: true, 71: true, 74: true, 75: true, 76: true, 80: true,
	93: true, 94: true, 95: true, 97: true,
}

func isWallSolid(id int) bool {
	return (id >= 0 && id < 48) || id == 53 || id == 54
}
func isWeapon(id int) bool             { return id == 85 || id == 86 }
func isAmmo(id int) bool               { return id == 84 }
func isScore(id int) bool              { return id == 87 || id == 88 || id == 89 }
func isDecorationSolid(id int) bool    { return typeIDsDecorationsSolid[id] }
func isDecorationNonSolid(id int) bool { return id >= 56 && id < 120 && !typeIDsDecorationsSolid[id] }
func isDoorHorizontal(id int) bool     { return id == 50 }
func isDoorVertical(id int) bool       { return id == 49 }
func isPlayerType(id int) bool         { return id == 255 }
func isEnemy(id int) bool              { return id == 130 || id == 131 }
